docs/markdown/Ad-Hoc Tools/adhoc-tool.md

```diff
@@ -0,0 +1,98 @@
+---
+title: "Integrating new tools without plugins"
+slug: "adhoc-tool"
+hidden: false
+createdAt: "2023-03-20T00:00:00.000Z"
+---
+
+# Integrating new tools without plugins
+
+The `adhoc_tool` target allows you to execute "runnable" targets inside the Pants sandbox. Runnable targets include first-party sources that can be run with `pants run`, 3rd-party dependencies like `python_requirement` or `jvm_artifact`, or even executables that exist on your system and managed externally to Pants.
+
+`adhoc_tool` provides you with the building blocks needed to put together a custom build process without needing to develop and maintain a plugin. The level of initial effort involved in using `adhoc_tool` is significantly lower than that of [writing a plugin](doc:plugins-overview), so it's well-suited to consuming one-off scripts, or for rapidly prototyping a process before actually writing a plugin. The tradeoff is that there is more manual work involved in defining build processes that reflect your codebase's structure, and that the targets that define the tools you consume are less easy to reuse.
+
+The `antlr` demo in the [`example-adhoc` respository](https://github.org/pantsbuild/example-adhoc) shows an example of running a JVM-based tool to transparently generate Python code that can be used in another language:
+
+```
+adhoc_tool(
+    name="run_antlr",
+    runnable=":antlr4",
+    args=["Expr.g4", "-Dlanguage=Python3", "-o", "expr_parser", "-package", "expr_parser",],
+    output_directories=["expr_parser",],
+    # These are consumed by `antlr`, but are not relevant to this target's dependents.
+    execution_dependencies=[":grammars"],
+    # These are needed by the code that is output by this target
+    output_dependencies=[":antlr4-python3-runtime",],
+    root_output_directory=".",
+    log_output=True,
+)
+```
+
+## `runnable` targets
+
+"Runnable" targets are targets that Pants knows how to execute within its sandbox. Generally, these correspond to targets that can be executed with the `pants run` goal, and include first-party source files, as well as third-party dependencies.
+
+The tool will be run with values from `args` specified as arguments. By default, the process' working directory will be the directory where the `BUILD` file is defined. This can be adjusted using the `workdir` field.
+
+When run by `adhoc_tool`, Pants assumes that the inputs provided to the process -- that is, the values of the `adhoc_tool`'s fields, and the contents of the dependencies -- fully describe the output. Output values will be [cached](doc:how-does-pants-work#caching) by Pants, and future invocations with identical inputs will be retrieved from the cache instead of being re-executed. If your process has behavior that is not fully defined by its inputs, Pants' behavior may be unexpected or inconsistent.
```

Comment:
Potentially:

```suggestion
When run by `adhoc_tool`, Pants assumes that the inputs provided to the process -- that is, the values of the `adhoc_tool`'s fields, and the contents of the runnable and execution dependencies -- fully describe the output. Output values will be [cached](doc:how-does-pants-work#caching) by Pants, and future invocations with identical inputs will be retrieved from the cache instead of being re-executed. If your process has behavior that is not fully defined by its inputs, Pants' behavior may be unexpected or inconsistent.
```

But maybe this is too much detail.